File: px-checkout/src/main/java/com/mercadopago/android/px/internal/datasource/CheckoutRepositoryImpl.java
package com.mercadopago.android.px.internal.datasource;

import android.os.Handler;
import android.os.HandlerThread;
import androidx.annotation.NonNull;
import com.mercadopago.android.px.configuration.AdvancedConfiguration;
import com.mercadopago.android.px.configuration.DiscountParamsConfiguration;
import com.mercadopago.android.px.configuration.PaymentConfiguration;
import com.mercadopago.android.px.internal.callbacks.MPCall;
import com.mercadopago.android.px.internal.features.FeatureProvider;
import com.mercadopago.android.px.internal.mappers.OneTapItemToDisabledPaymentMethodMapper;
import com.mercadopago.android.px.internal.repository.AmountConfigurationRepository;
import com.mercadopago.android.px.internal.repository.CheckoutRepository;
import com.mercadopago.android.px.internal.repository.DisabledPaymentMethodRepository;
import com.mercadopago.android.px.internal.repository.DiscountRepository;
import com.mercadopago.android.px.internal.repository.ExperimentsRepository;
import com.mercadopago.android.px.internal.repository.ModalRepository;
import com.mercadopago.android.px.internal.repository.OneTapItemRepository;
import com.mercadopago.android.px.internal.repository.PayerComplianceRepository;
import com.mercadopago.android.px.internal.repository.PayerPaymentMethodKey;
import com.mercadopago.android.px.internal.repository.PayerPaymentMethodRepository;
import com.mercadopago.android.px.internal.repository.PaymentMethodRepository;
import com.mercadopago.android.px.internal.repository.PaymentSettingRepository;
import com.mercadopago.android.px.internal.services.CheckoutService;
import com.mercadopago.android.px.internal.tracking.TrackingRepository;
import com.mercadopago.android.px.internal.util.JsonUtil;
import com.mercadopago.android.px.model.exceptions.ApiException;
import com.mercadopago.android.px.model.internal.CheckoutResponse;
import com.mercadopago.android.px.model.internal.DisabledPaymentMethod;
import com.mercadopago.android.px.model.internal.InitRequest;
import com.mercadopago.android.px.model.internal.OneTapItem;
import com.mercadopago.android.px.preferences.CheckoutPreference;
import com.mercadopago.android.px.services.Callback;
import com.mercadopago.android.px.tracking.internal.MPTracker;
import java.util.List;
import java.util.Map;

public class CheckoutRepositoryImpl implements CheckoutRepository {

    private static final int MAX_REFRESH_RETRIES = 4;
    private static final int DEFAULT_RETRY_DELAY = 500;
    private static final int LONG_RETRY_DELAY = 5000;

    @NonNull private final CheckoutService checkoutService;
    @NonNull /* default */ final PaymentSettingRepository paymentSettingRepository;
    @NonNull /* default */ final ExperimentsRepository experimentsRepository;
    @NonNull /* default */ DisabledPaymentMethodRepository disabledPaymentMethodRepository;
    @NonNull private final MPTracker tracker;
    @NonNull private final PayerPaymentMethodRepository payerPaymentMethodRepository;
    @NonNull private final OneTapItemRepository oneTapItemRepository;
    @NonNull private final PaymentMethodRepository paymentMethodRepository;
    @NonNull private final ModalRepository modalRepository;
    @NonNull private final PayerComplianceRepository payerComplianceRepository;
    @NonNull private final AmountConfigurationRepository amountConfigurationRepository;
    @NonNull private final DiscountRepository discountRepository;
    @NonNull private final TrackingRepository trackingRepository;
    @NonNull private final CardStatusRepository cardStatusRepository;
    @NonNull private final FeatureProvider featureProvider;
    /* default */ int refreshRetriesAvailable = MAX_REFRESH_RETRIES;
    /* default */ Handler retryHandler;

    public CheckoutRepositoryImpl(@NonNull final PaymentSettingRepository paymentSettingRepository,
        @NonNull final ExperimentsRepository experimentsRepository,
        @NonNull final DisabledPaymentMethodRepository disabledPaymentMethodRepository,
        @NonNull final CheckoutService checkoutService,
        @NonNull final TrackingRepository trackingRepository, @NonNull final MPTracker tracker,
        @NonNull final PayerPaymentMethodRepository payerPaymentMethodRepository,
        @NonNull final OneTapItemRepository oneTapItemRepository,
        @NonNull final PaymentMethodRepository paymentMethodRepository,
        @NonNull final ModalRepository modalRepository,
        @NonNull final PayerComplianceRepository payerComplianceRepository,
        @NonNull final AmountConfigurationRepository amountConfigurationRepository,
        @NonNull final DiscountRepository discountRepository,
        @NonNull final CardStatusRepository cardStatusRepository,
        @NonNull final FeatureProvider featureProvider) {
        this.paymentSettingRepository = paymentSettingRepository;
        this.experimentsRepository = experimentsRepository;
        this.disabledPaymentMethodRepository = disabledPaymentMethodRepository;
        this.checkoutService = checkoutService;
        this.trackingRepository = trackingRepository;
        this.tracker = tracker;
        this.payerPaymentMethodRepository = payerPaymentMethodRepository;
        this.oneTapItemRepository = oneTapItemRepository;
        this.paymentMethodRepository = paymentMethodRepository;
        this.modalRepository = modalRepository;
        this.payerComplianceRepository = payerComplianceRepository;
        this.amountConfigurationRepository = amountConfigurationRepository;
        this.discountRepository = discountRepository;
        this.cardStatusRepository = cardStatusRepository;
        this.featureProvider = featureProvider;
    }

    @NonNull
    @Override
    public MPCall<CheckoutResponse> checkout() {
        return newCall(getPostResponse());
    }

    protected void configure(@NonNull final CheckoutResponse checkoutResponse) {
        if (checkoutResponse.getPreference() != null) {
            paymentSettingRepository.configure(checkoutResponse.getPreference());
        }
        paymentSettingRepository.configure(checkoutResponse.getSite());
        paymentSettingRepository.configure(checkoutResponse.getCurrency());
        paymentSettingRepository.configure(checkoutResponse.getConfiguration());
        experimentsRepository.configure(checkoutResponse.getExperiments());
        payerPaymentMethodRepository.configure(checkoutResponse.getPayerPaymentMethods());
        oneTapItemRepository.configure(checkoutResponse.getOneTapItems());
        paymentMethodRepository.configure(checkoutResponse.getAvailablePaymentMethods());
        modalRepository.configure(checkoutResponse.getModals());
        payerComplianceRepository.configure(checkoutResponse.getPayerCompliance());
        amountConfigurationRepository.configure(checkoutResponse.getDefaultAmountConfiguration());
        discountRepository.configure(checkoutResponse.getDiscountsConfigurations());
        disabledPaymentMethodRepository.configure(
            new OneTapItemToDisabledPaymentMethodMapper().map(checkoutResponse.getOneTapItems()));

        tracker.setExperiments(experimentsRepository.getExperiments());
    }

    @Override
    public void lazyConfigure(@NonNull final CheckoutResponse checkoutResponse) {
        getPostResponse().call(checkoutResponse);
    }

    interface PostResponse {
        void call(CheckoutResponse checkoutResponse);
    }

    /* default */ PostResponse getPostResponse() {
        return this::configure;
    }

    private PostResponse noPostResponse() {
        return initResponse -> {
        };
    }

    @NonNull
    private MPCall<CheckoutResponse> newCall(@NonNull final PostResponse postResponse) {
        return new MPCall<CheckoutResponse>() {

            @Override
            public void enqueue(final Callback<CheckoutResponse> callback) {
                newRequest().enqueue(getInternalCallback(callback));
            }

            @NonNull /* default */ Callback<CheckoutResponse> getInternalCallback(
                final Callback<CheckoutResponse> callback) {
                return new Callback<CheckoutResponse>() {
                    @Override
                    public void success(final CheckoutResponse checkoutResponse) {
                        postResponse.call(checkoutResponse);
                        callback.success(checkoutResponse);
                    }

                    @Override
                    public void failure(final ApiException apiException) {
                        callback.failure(apiException);
                    }
                };
            }
        };
    }

    @NonNull /* default */ MPCall<CheckoutResponse> newRequest() {
        final CheckoutPreference checkoutPreference = paymentSettingRepository.getCheckoutPreference();
        final PaymentConfiguration paymentConfiguration = paymentSettingRepository.getPaymentConfiguration();

        final AdvancedConfiguration advancedConfiguration = paymentSettingRepository.getAdvancedConfiguration();
        final DiscountParamsConfiguration discountParamsConfiguration =
            advancedConfiguration.getDiscountParamsConfiguration();

        final Map<String, Object> body = JsonUtil.getMapFromObject(
            new InitRequest.Builder(paymentSettingRepository.getPublicKey())
                .setCharges(paymentConfiguration.getCharges())
                .setDiscountParamsConfiguration(discountParamsConfiguration)
                .setCheckoutFeatures(featureProvider.getAvailableFeatures())
                .setCheckoutPreference(checkoutPreference)
                .setFlow(trackingRepository.getFlowId())
                .setCardsStatus(cardStatusRepository.getCardsStatus())
                .build());

        final String preferenceId = paymentSettingRepository.getCheckoutPreferenceId();
        if (preferenceId != null) {
            return checkoutService.checkout(preferenceId, paymentSettingRepository.getPrivateKey(), body);
        } else {
            return checkoutService.checkout(paymentSettingRepository.getPrivateKey(), body);
        }
    }

    @NonNull
    @Override
    public MPCall<CheckoutResponse> refreshWithNewCard(@NonNull final String cardId) {
        return callback -> newCall(noPostResponse()).enqueue(getRefreshWithNewCardCallback(cardId, callback));
    }

    /* default */ Callback<CheckoutResponse> getRefreshWithNewCardCallback(@NonNull final String cardId,
        @NonNull final Callback<CheckoutResponse> callback) {
        final Map<PayerPaymentMethodKey, DisabledPaymentMethod> disabledPaymentMethodMap =
            disabledPaymentMethodRepository.getValue();
        return new Callback<CheckoutResponse>() {
            @Override
            public void success(final CheckoutResponse checkoutResponse) {
                final boolean retryAvailable = --refreshRetriesAvailable > 0;
                final List<OneTapItem> oneTap = checkoutResponse.getOneTapItems();
                boolean cardFound = false;
                boolean retryNeeded = false;
                for (final OneTapItem node : oneTap) {
                    if (node.isCard() && node.getCard().getId().equals(cardId)) {
                        cardFound = true;
                        retryNeeded = node.getCard().getRetry().isNeeded();
                        break;
                    }
                }
                if (cardFound && (!retryNeeded || !retryAvailable)) {
                    refreshRetriesAvailable = MAX_REFRESH_RETRIES;
                    new OneTapItemSorter(oneTap, disabledPaymentMethodMap)
                        .setPrioritizedCardId(cardId).sort();
                    getPostResponse().call(checkoutResponse);
                    callback.success(checkoutResponse);
                } else if (retryAvailable) {
                    final int retryDelay = retryNeeded ? LONG_RETRY_DELAY : DEFAULT_RETRY_DELAY;
                    if (retryHandler == null) {
                        final HandlerThread thread = new HandlerThread("MyInitRetryThread");
                        thread.start();
                        retryHandler = new Handler(thread.getLooper());
                    }
                    retryHandler.postDelayed(() -> refreshWithNewCard(cardId).enqueue(callback), retryDelay);
                } else {
                    callback.failure(new ApiException());
                }
            }

            @Override
            public void failure(final ApiException apiException) {
                callback.failure(apiException);
            }
        };
    }
}

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/datasource/CheckoutRepositoryImpl.kt
package com.mercadopago.android.px.internal.datasource

import com.mercadopago.android.px.internal.adapters.NetworkApi
import com.mercadopago.android.px.internal.callbacks.ApiResponse
import com.mercadopago.android.px.internal.mappers.CustomChargeToPaymentTypeChargeMapper
import com.mercadopago.android.px.internal.callbacks.Response
import com.mercadopago.android.px.internal.mappers.InitRequestBodyMapper
import com.mercadopago.android.px.internal.mappers.OneTapItemToDisabledPaymentMethodMapper
import com.mercadopago.android.px.internal.repository.*
import com.mercadopago.android.px.internal.services.CheckoutService
import com.mercadopago.android.px.internal.util.ApiUtil
import com.mercadopago.android.px.model.exceptions.ApiException
import com.mercadopago.android.px.model.exceptions.MercadoPagoError
import com.mercadopago.android.px.model.internal.CheckoutResponse
import com.mercadopago.android.px.tracking.internal.MPTracker
import kotlinx.coroutines.delay

internal typealias ApiResponseCallback<R> = ApiResponse<R, ApiException>
internal typealias ResponseCallback<R> = Response<R, MercadoPagoError>

internal open class CheckoutRepositoryImpl(
    private val paymentSettingRepository: PaymentSettingRepository,
    private val experimentsRepository: ExperimentsRepository,
    private val disabledPaymentMethodRepository: DisabledPaymentMethodRepository,
    private val networkApi: NetworkApi,
    private val tracker: MPTracker,
    private val payerPaymentMethodRepository: PayerPaymentMethodRepository,
    private val oneTapItemRepository: OneTapItemRepository,
    private val paymentMethodRepository: PaymentMethodRepository,
    private val modalRepository: ModalRepository,
    private val payerComplianceRepository: PayerComplianceRepository,
    private val amountConfigurationRepository: AmountConfigurationRepository,
    private val discountRepository: DiscountRepository,
    private val customChargeToPaymentTypeChargeMapper: CustomChargeToPaymentTypeChargeMapper,
    private val initRequestBodyMapper: InitRequestBodyMapper,
    private val oneTapItemToDisabledPaymentMethodMapper: OneTapItemToDisabledPaymentMethodMapper
) : CheckoutRepository {

    override suspend fun checkout() = doCheckout(null)

    override fun configure(checkoutResponse: CheckoutResponse) {
        if (checkoutResponse.preference != null) {
            paymentSettingRepository.configure(checkoutResponse.preference)
        }
        paymentSettingRepository.configure(checkoutResponse.site)
        paymentSettingRepository.configure(checkoutResponse.currency)
        paymentSettingRepository.configure(checkoutResponse.configuration)

        // TODO: Remove null check when backend has IDC ready
        checkoutResponse.customCharges?.let {
            paymentSettingRepository.configure(
                customChargeToPaymentTypeChargeMapper.map(it)
            )
        }

        experimentsRepository.configure(checkoutResponse.experiments)
        payerPaymentMethodRepository.configure(checkoutResponse.payerPaymentMethods)
        oneTapItemRepository.configure(checkoutResponse.oneTapItems)
        paymentMethodRepository.configure(checkoutResponse.availablePaymentMethods)
        modalRepository.configure(checkoutResponse.modals)
        payerComplianceRepository.configure(checkoutResponse.payerCompliance)
        amountConfigurationRepository.configure(checkoutResponse.defaultAmountConfiguration)
        discountRepository.configure(checkoutResponse.discountsConfigurations)
        disabledPaymentMethodRepository.configure(
            oneTapItemToDisabledPaymentMethodMapper.map(checkoutResponse.oneTapItems)
        )
        tracker.setExperiments(experimentsRepository.experiments)
    }

    override suspend fun checkoutWithNewCard(cardId: String): Response<CheckoutResponse, MercadoPagoError> {
        var retriesAvailable = MAX_REFRESH_RETRIES
        var findCardResult = checkoutAndFindCard(cardId)
        var lastSuccessResponse = findCardResult.response.takeIf { it is Response.Success }
        while (findCardResult.retryNeeded && retriesAvailable > 0) {
            retriesAvailable--
            delay(RETRY_DELAY)
            findCardResult = checkoutAndFindCard(cardId)
            if (findCardResult.response is Response.Success) {
                lastSuccessResponse = findCardResult.response
            }
        }

        return lastSuccessResponse ?: findCardResult.response
    }

    private suspend fun doCheckout(cardId: String?): ResponseCallback<CheckoutResponse> {
        val body = initRequestBodyMapper.map(paymentSettingRepository, cardId)
        val preferenceId = paymentSettingRepository.checkoutPreferenceId
        val apiResponse = networkApi.apiCallForResponse(CheckoutService::class.java) {
            if (preferenceId != null) {
                it.checkout(preferenceId, body)
            } else {
                it.checkout(body)
            }
        }
        return when (apiResponse) {
            is ApiResponse.Failure -> Response.Failure(
                MercadoPagoError(
                    apiResponse.exception,
                    ApiUtil.RequestOrigin.POST_INIT
                )
            )
            is ApiResponse.Success -> Response.Success(apiResponse.result)
        }
    }

    private suspend fun checkoutAndFindCard(cardId: String): FindCardResult {
        return when (val response = doCheckout(cardId)) {
            is Response.Success -> findCard(response, cardId)
            is Response.Failure -> FindCardResult(false, response)
        }
    }

    private fun findCard(response: Response.Success<CheckoutResponse>, cardId: String): FindCardResult {
        var retryNeeded = false
        for (node in response.result.oneTapItems) {
            if (node.isCard && node.card.id == cardId) {
                retryNeeded = node.card.retry.isNeeded
                break
            }
        }
        return FindCardResult(retryNeeded, response)
    }

    private data class FindCardResult(
        val retryNeeded: Boolean,
        var response: ResponseCallback<CheckoutResponse>
    )

    companion object {
        private const val MAX_REFRESH_RETRIES = 3
        private const val RETRY_DELAY = 5000L
    }
}
